pilot/pkg/model/push_context.go

```diff
@@ -979,6 +1052,44 @@ func (ps *PushContext) initDestinationRules(env *Environment) error {
 	return nil
 }
 
+// AuthenticationPolicyForWorkload returns the matching auth policy for a given service
+// This replaces store.AuthenticationPolicyForWorkload
+func (ps *PushContext) AuthenticationPolicyForWorkload(service *Service, port *Port) *authn.Policy {
+	var workloadPolicy *authn.Policy
+	// Match by Service hostname
+	if workloadPolicy = authenticationPolicyForWorkload(ps.AuthnPolicies.policies[service.Hostname], port); workloadPolicy != nil {
+		return workloadPolicy
+	}
+
+	// Match by namespace
+	if workloadPolicy = authenticationPolicyForWorkload(ps.AuthnPolicies.policies[host.Name(service.Attributes.Namespace)], port); workloadPolicy != nil {
+		return workloadPolicy
+	}
+
+	// Use default global authentication policy if no others found
+	return ps.AuthnPolicies.defaultMeshPolicy
+}
+
+func authenticationPolicyForWorkload(policiesByPort []*authnPolicyByPort, port *Port) *authn.Policy {
+	var matchedPolicy *authn.Policy
+	if policiesByPort == nil {
+		return nil
+	}
+
+	for i, policyByPort := range policiesByPort {
+		if policyByPort.portSelector == nil && matchedPolicy == nil {
+			matchedPolicy = policiesByPort[i].policy
+		}
+
+		if port.Match(policyByPort.portSelector) {
```

Comment:
after handling line 1080, we're still invoking this line, and then what's the behavior to invoke `port.Match(nil)`? return true? Then we entered block below and break the loop, which is wrong.

Can you confirm? If so, have a unit test for this?

Reply:
@incfly from `pilot/pkg/model/service.go`
```
func (port Port) Match(portSelector *authn.PortSelector) bool {
	if portSelector == nil {
		return true
	}
```
>port.Match(nil)? return true? Then we entered block below and break the loop, which is wrong.

why is this wrong?  No `Port_Selector` was specified in the `authn.Policy`, so why shouldn't this count as a match for the service/port combination passed in to `AuthenticationPolicyForWorkload(service *Service, port *Port)`

Or are you questioning the validity of the 1st check on line 1080.  I added that because I wasn't sure if there would ever be a scenario where `port` could be nil

Reply:
Because if I understand correctly,

the intended authn/z api semantics is the most specific API wins, the policy with port takes higher priority compared with policy without port selector. @diemtvu please correct me if I am wrong.

The implementation right now is the first matched one used.

Reply:
I take another look in the existing implementation, https://github.com/istio/istio/blob/master/pilot/pkg/model/config.go#L549

So there seems like we're not doing the port selector specified takes higher priority either... So this is not changing status quo.

Resolving.